components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java

```diff
@@ -0,0 +1,349 @@
+                        //in the case of email, there will be more than one '@'
+                        boolean isEmailUsernameEnabled = Boolean.parseBoolean(CarbonUtils.getServerConfiguration().
+                                getFirstProperty("EnableEmailUserName"));
+                        if (isEmailUsernameEnabled || (username.endsWith(SUPER_TENANT_SUFFIX) && count <= 1)) {
+                            username = MultitenantUtils.getTenantAwareUsername(username);
+                        }
+                    }
+                    if (log.isDebugEnabled()) {
+                        log.debug("username = " + username);
+                    }
+                    tenantId = realmService.getTenantManager().getTenantId(tenantDomain);
+                    carbonContext.setTenantDomain(tenantDomain);
+                    carbonContext.setTenantId(tenantId);
+                    carbonContext.setUsername(username);
+                    if (!tenantDomain.equals(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)) {
+                        APIUtil.loadTenantConfigBlockingMode(tenantDomain);
+                    }
+                    return true;
+                } catch (UserStoreException e) {
+                    log.error("Error while retrieving tenant id for tenant domain: " + tenantDomain, e);
+                }
+                log.info("Scope validation success");
+                return true;
+            }
+            log.error("scopes validation failed");
+            return false;
+        }
+        log.error("scopes validation failed");
+        return false;
+    }
+
+    /**
+     * Get signed jwt info.
+     *
+     * @param accessToken JWT token
+     * @return SignedJWTInfo : Signed token info
+     */
+    @MethodStats
+    private SignedJWTInfo getSignedJwt(String accessToken) throws ParseException {
+
+        SignedJWT signedJWT = SignedJWT.parse(accessToken);
+        JWTClaimsSet jwtClaimsSet = signedJWT.getJWTClaimsSet();
+        return new SignedJWTInfo(accessToken, signedJWT, jwtClaimsSet);
+    }
+
+    /**
+     * Validate the JWT token.
+     *
+     * @param jti           jwtTokenIdentifier
+     * @param signedJWTInfo signed jwt info object
+     * @return JWTValidationInfo : token validated info
+     */
+    @MethodStats
+    private JWTValidationInfo validateJWTToken(SignedJWTInfo signedJWTInfo, String jti, String accessToken,
+                                               String jwtHeader, URL basePath) throws APIManagementException {
+
+        JWTValidationInfo jwtValidationInfo;
+        String issuer = signedJWTInfo.getJwtClaimsSet().getIssuer();
+
+        if (StringUtils.isNotEmpty(issuer)) {
+            //validate Issuer
+            List<String> audiences = signedJWTInfo.getJwtClaimsSet().getAudience();
+            if (tokenIssuers != null && tokenIssuers.containsKey(issuer)) {
+                jwtValidator.loadTokenIssuerConfiguration(tokenIssuers.get(issuer));
```

Comment:
Why do we need loadTokenIssuerConfiguration during token validation path? Shouldn't we move this to the constructor?

Reply:
Token issuer configs should be loaded according to the issuer of the current token. So we get the issuer by getting signed JWT info. getting signed JWT info is happening in authenticate method level and not at the constructor level. So it is possible to move this inside authenticate method other than keeping here. What would be the best ?